crates/forge/src/lib.rs

```diff
@@ -285,146 +288,116 @@ fn try_close_tmp_dir(maybe_tmp_dir: Option<TempDir>) -> Result<()> {
+    let runner = Arc::new(
+        SierraCasmRunner::new(
+            sierra_program.clone(),
+            Some(MetadataComputationConfig::default()),
+            OrderedHashMap::default(),
+        )
+        .context("Failed setting up runner.")?,
+    );
 
     let mut was_fuzzed = false;
-    let mut results = vec![];
+    let mut tasks = FuturesOrdered::new();
+    let test_cases = &tests.test_cases;
 
-    for (i, case) in tests.test_cases.iter().enumerate() {
+    for case in test_cases.iter() {
         let case_name = case.name.as_str();
+
         let function = runner.find_function(case_name)?;
         let args = function_args(function, &BUILTINS);
 
-        let result = if args.is_empty() {
-            let result = run_from_test_case(
-                package_root,
-                &runner,
-                case,
-                runner_config.fork_targets.as_ref(),
-                &runner_params.contracts,
-                &runner_params.predeployed_contracts,
-                vec![],
-                &runner_params.environment_variables,
-            )?;
-            pretty_printing::print_test_result(&result, None);
-
-            result
-        } else {
-            was_fuzzed = true;
-            let (result, runs) = run_with_fuzzing(
-                package_root,
-                runner_config,
-                runner_params,
-                &runner,
-                case,
-                &args,
-            )?;
-            pretty_printing::print_test_result(&result, Some(runs));
-
-            result
-        };
+        let case = Arc::new(case.clone());
+        let runner_config = runner_config.clone();
+        let args: Vec<ConcreteTypeId> = args.into_iter().cloned().collect();
+        let runner = runner.clone();
 
-        results.push(result.clone());
+        tasks.push_back(choose_test_strategy_and_run(
+            case.clone(),
+            runner,
+            runner_config.clone(),
+            args,
+        ));
+    }
 
-        if runner_config.exit_first {
-            if let TestCaseSummary::Failed { .. } = result {
-                for case in &tests.test_cases[i + 1..] {
-                    let skipped_result = TestCaseSummary::skipped(case);
-                    pretty_printing::print_test_result(&skipped_result, None);
-                    results.push(skipped_result);
-                }
-                return Ok((
-                    TestCrateSummary {
-                        test_case_summaries: results,
-                        runner_exit_status: RunnerStatus::TestFailed,
-                        test_crate_type: tests.test_crate_type,
-                    },
-                    was_fuzzed,
-                ));
-            }
+    let mut results = vec![];
+
+    while let Some(task) = tasks.next().await {
+        let await_task = task??;
+
+        let result = await_task;
+
+        if result.runs().is_some() {
+            was_fuzzed = true;
```

Comment:
I'm not sure if we should be detecting if the task was fuzzed like this. This kind of looks like a thing we should just be aware of here in code.

Reply:
#841, out of scope of this PR